src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/WorkspaceSemanticTokensRefreshNotifier.cs

```diff
@@ -0,0 +1,113 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Razor.Workspaces.Protocol;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer;
+
+internal class WorkspaceSemanticTokensRefreshNotifier : IWorkspaceSemanticTokensRefreshNotifier, IDisposable
+{
+    private static readonly TimeSpan s_delay = TimeSpan.FromMilliseconds(250);
+
+    private readonly IClientCapabilitiesService _clientCapabilitiesService;
+    private readonly IClientConnection _clientConnection;
+    private readonly CancellationTokenSource _disposeCancellationSource;
+    private readonly IDisposable _optionsChangeListener;
+
+    private readonly object _gate = new();
+    private bool _waitingToRefresh;
+    private Task _refreshTask = Task.CompletedTask;
```

Comment:
I'm unfamiliar with `TaskDelayScheduler`, so it didn't occur to me. But I'll take a look -- thanks!

Reply:
After taking a look, I think I'll stick with the simpler implementation here. `TaskDelayScheduler` uses a fire-and-forget `Task` with no cancellation token. Also, it doesn't return the `Task`, which would make it more awkward for tests to wait for the delay to complete. It might be good to revisit `TaskDelayScheduler`, but I think that's out-of-scope for this change. Is that OK with you?

Reply:
100%